### ISSUE TITLE
Benchmark package managers

### ISSUE THREAD

**Issue by joshhunt** [grafana/grafana#77893]
Benchmark various frontend package managers to understand how they perform with the Grafana repo.

**Comment by joshhunt** [comment]
Did some rough benchmarks on my Macbook Pro w/M1 Pro processor.


tl;dr:
 - Yarn pnp (current setup) is fastest
 - Yarn pnpm is the slowest
 - pnpm sits in the middle

Worth noting that I also ran the benchmarks on my PC within WSL 2 Ubuntu and got pretty different results - pnpm was the fastest option for a few of these. I have not ran the benchmarks on 'native' x64 linux.


### No Caches

Simulates a fresh checkout of Grafana with a clear global package manager cache.

yarn pnp-linker ran

- 1.19 times faster than pnpm
- 2.28 times faster than yarn pnpm-linker

| Command            |       Mean [s] | Min [s] | Max [s] |    Relative |
| :----------------- | -------------: | ------: | ------: | ----------: |
| `yarn pnp-linker`  | 18.629 ± 0.517 |  17.875 |  19.451 |        1.00 |
| `yarn pnpm-linker` | 42.425 ± 0.433 |  41.784 |  43.068 | 2.28 ± 0.07 |
| `pnpm`             | 22.198 ± 0.821 |  21.291 |  23.817 | 1.19 ± 0.06 |

<details>
<summary>See command and raw output</summary>

<pre>
hyperfine --runs 10 --warmup 2
    --prepare 'git reset HEAD --hard; clear-all-pm-caches.sh; git apply apply-yarn-pnp.patch || true' 'yarn install; echo yarn-pnp'
    --prepare 'git reset HEAD --hard; clear-all-pm-caches.sh' 'yarn install; echo yarn-pnpm'
    --prepare 'git reset HEAD --hard; clear-all-pm-caches.sh' 'pnpm install; echo pnp'
    --cleanup 'git reset HEAD --hard'
    --export-markdown results-no-caches_2023-11-08-15-08-31.md

Benchmark 1: yarn pnp-linker
  Time (mean ± σ):     18.629 s ±  0.517 s    [User: 58.540 s, System: 5.433 s]
  Range (min … max):   17.875 s … 19.451 s    10 runs

Benchmark 2: yarn pnpm-linker
  Time (mean ± σ):     42.425 s ±  0.433 s    [User: 76.243 s, System: 69.034 s]
  Range (min … max):   41.784 s … 43.068 s    10 runs

Benchmark 3: pnpm
  Time (mean ± σ):     22.198 s ±  0.821 s    [User: 17.863 s, System: 41.340 s]
  Range (min … max):   21.291 s … 23.817 s    10 runs

Summary
  yarn pnp-linker ran
    1.19 ± 0.06 times faster than pnpm
    2.28 ± 0.07 times faster than yarn pnpm-linker
</pre>
</details>

### Warm cache, empty node_modules

This doesn't typically come up in regular scenarios.

Yarn pnp-linker strongly excels here because this is more or less a no-op.

yarn pnp-linker ran

- 6.52 times faster than pnpm
- 17.68 times faster than yarn pnpm-linker

| Command            |       Mean [s] | Min [s] | Max [s] |     Relative |
| :----------------- | -------------: | ------: | ------: | -----------: |
| `yarn pnp-linker`  |  1.391 ± 0.100 |   1.239 |   1.467 |         1.00 |
| `yarn pnpm-linker` | 24.580 ± 0.117 |  24.387 |  24.747 | 17.68 ± 1.27 |
| `pnpm`             |  9.072 ± 0.070 |   8.985 |   9.216 |  6.52 ± 0.47 |

<details>
<summary>See command and raw output</summary>

<pre>
hyperfine --runs 10 --warmup 2
    --prepare 'git reset HEAD --hard; clear-node-modules.sh; git apply apply-yarn-pnp.patch || true' 'yarn install; echo yarn-pnp'
    --prepare 'git reset HEAD --hard; clear-node-modules.sh' 'yarn install; echo yarn-pnpm'
    --prepare 'git reset HEAD --hard; clear-node-modules.sh' 'pnpm install; echo pnp'
    --cleanup 'git reset HEAD --hard'
    --export-markdown results-warm-caches_2023-11-08-15-30-42.md

Benchmark 1: yarn pnp-linker
  Time (mean ± σ):      1.391 s ±  0.100 s    [User: 2.289 s, System: 0.231 s]
  Range (min … max):    1.239 s …  1.467 s    10 runs

Benchmark 2: yarn pnpm-linker
  Time (mean ± σ):     24.580 s ±  0.117 s    [User: 19.349 s, System: 65.352 s]
  Range (min … max):   24.387 s … 24.747 s    10 runs

Benchmark 3: pnpm
  Time (mean ± σ):      9.072 s ±  0.070 s    [User: 7.769 s, System: 25.227 s]
  Range (min … max):    8.985 s …  9.216 s    10 runs

Summary
  yarn pnp-linker ran
    6.52 ± 0.47 times faster than pnpm
   17.68 ± 1.27 times faster than yarn pnpm-linker
</pre>
</details>

### Up to date

The reflexive `yarn install` I do quite frequently.

Less required with pnpm because it tolerates running with outdated dependencies (for better or worsed).

yarn pnp-linker ran

- 2.11 times faster than pnpm
- 6.08 times faster than yarn pnpm-linker

| Command            |      Mean [s] | Min [s] | Max [s] |    Relative |
| :----------------- | ------------: | ------: | ------: | ----------: |
| `yarn pnp-linker`  | 1.448 ± 0.017 |   1.433 |   1.474 |        1.00 |
| `yarn pnpm-linker` | 8.805 ± 0.040 |   8.768 |   8.912 | 6.08 ± 0.08 |
| `pnpm`             | 3.058 ± 0.026 |   3.017 |   3.122 | 2.11 ± 0.03 |



<details>
<summary>See command and raw output</summary>
hyperfine --runs 10 --warmup 3
  --prepare 'git reset HEAD --hard; git apply /Users/joshhunt/dev/gf-pm-bench/apply-yarn-pnp.patch || true' 'yarn install; echo yarn-pnp'
  --prepare 'git reset HEAD --hard' 'yarn install; echo yarn-pnpm'
  --prepare 'git reset HEAD --hard' 'pnpm install; echo pnp'
  --cleanup 'git reset HEAD --hard; rm -rf node_modules'
  --export-markdown /Users/joshhunt/dev/gf-pm-bench/results-up-to-date_2023-11-08-16-25-19.md

Benchmark 1: yarn pnp-linker
Time (mean ± σ): 1.448 s ± 0.017 s [User: 2.322 s, System: 0.251 s]
Range (min … max): 1.433 s … 1.474 s 10 runs

Benchmark 2: yarn pnpm-linker
Time (mean ± σ): 8.805 s ± 0.040 s [User: 13.527 s, System: 6.117 s]
Range (min … max): 8.768 s … 8.912 s 10 runs

Benchmark 3: pnpm
Time (mean ± σ): 3.058 s ± 0.026 s [User: 3.821 s, System: 1.067 s]
Range (min … max): 3.017 s … 3.122 s 10 runs

Summary
yarn pnp-linker ran
2.11 ± 0.03 times faster than pnpm
6.08 ± 0.08 times faster than yarn pnpm-linker

</pre>
</details>

### Updating deps

Simulates pulling main and installing 15 dependencies have been updated over the past few days. See contents of `new-deps.patch`` for details

yarn pnp-linker ran

- 1.70 times faster than pnpm
- 4.66 times faster than yarn pnpm-linker


| Command            |       Mean [s] | Min [s] | Max [s] |    Relative |
| :----------------- | -------------: | ------: | ------: | ----------: |
| `yarn pnp-linker`  |  2.231 ± 0.117 |   2.054 |   2.424 |        1.00 |
| `yarn pnpm-linker` | 10.389 ± 0.350 |  10.032 |  11.226 | 4.66 ± 0.29 |
| `pnpm`             |  3.784 ± 0.224 |   3.263 |   4.114 | 1.70 ± 0.13 |


<details>
<summary>See command and raw output</summary>
hyperfine --runs 10 --warmup 2
    --prepare 'git reset HEAD --hard; git apply apply-yarn-pnp.patch || true; yarn cache clean --all; yarn install; git apply new-deps.patch || true' 'yarn install; echo yarn-pnp'
    --prepare 'git reset HEAD --hard; yarn cache clean --all; yarn install; git apply new-deps.patch || true' 'yarn install; echo yarn-pnpm'
    --prepare 'git reset HEAD --hard; rm -rf /Users/joshhunt/Library/pnpm/store/v3; pnpm install; git apply new-deps.patch || true' 'pnpm install; echo pnp'
    --cleanup 'git reset HEAD --hard'
    --export-markdown results-update-deps_2023-11-08-15-41-31.md

Benchmark 1: yarn pnp-linker
Time (mean ± σ): 2.231 s ± 0.117 s [User: 4.256 s, System: 0.412 s]
Range (min … max): 2.054 s … 2.424 s 10 runs

Benchmark 2: yarn pnpm-linker
Time (mean ± σ): 10.389 s ± 0.350 s [User: 16.243 s, System: 8.973 s]
Range (min … max): 10.032 s … 11.226 s 10 runs

Benchmark 3: pnpm
Time (mean ± σ): 3.784 s ± 0.224 s [User: 4.964 s, System: 1.880 s]
Range (min … max): 3.263 s … 4.114 s 10 runs

Summary
yarn pnp-linker ran
1.70 ± 0.13 times faster than pnpm
4.66 ± 0.29 times faster than yarn pnpm-linker

</pre>
</details>

**Comment by vishalnandagopal** [comment]
Any chance of checking the speeds with bun.sh, using it for just performing `bun install`?
